source/extensions/filters/http/header_mutation/header_mutation.cc

```diff
@@ -12,70 +12,154 @@ namespace Extensions {
   // first, then virtual host, then per route).
-  route_configs_ = Http::Utility::getAllPerFilterConfig<PerRouteHeaderMutation>(decoder_callbacks_);
-
+  route_configs_ = Http::Utility::getAllPerFilterConfig<PerRouteHeaderMutation>(callbacks);
+
+  // The order of returned route configs is route table first, then virtual host, then
+  // per route. That means the per route configs will be evaluated at the end and will
+  // override the more general virtual host and route table configs.
+  //
+  // So, if most_specific_header_mutations_wins is false, we need to reverse the order
+  // of route configs.
   if (!config_->mostSpecificHeaderMutationsWins()) {
-    // most_specific_wins means that most specific level per filter config is evaluated last. In
-    // other words, header mutations are evaluated in ascending order of specificity (same order as
-    // `getAllPerFilterConfig` above returns).
-    // Thus, here we reverse iterate the vector when `most_specific_wins` is false.
-    for (auto it = route_configs_.rbegin(); it != route_configs_.rend(); ++it) {
-      (*it).get().mutations().mutateRequestHeaders(headers, ctx, decoder_callbacks_->streamInfo());
-    }
-  } else {
-    for (const PerRouteHeaderMutation& route_config : route_configs_) {
-      route_config.mutations().mutateRequestHeaders(headers, ctx, decoder_callbacks_->streamInfo());
-    }
+    std::reverse(route_configs_.begin(), route_configs_.end());
+  }
+}
+
+Http::FilterHeadersStatus HeaderMutation::decodeHeaders(Http::RequestHeaderMap& headers, bool) {
+  Formatter::HttpFormatterContext context{&headers};
+  config_->mutations().mutateRequestHeaders(headers, context, decoder_callbacks_->streamInfo());
+
+  maybeInitializeRouteConfigs(decoder_callbacks_);
+
+  for (const PerRouteHeaderMutation& route_config : route_configs_) {
+    route_config.mutations().mutateRequestHeaders(headers, context,
+                                                  decoder_callbacks_->streamInfo());
   }
 
   return Http::FilterHeadersStatus::Continue;
 }
 
 Http::FilterHeadersStatus HeaderMutation::encodeHeaders(Http::ResponseHeaderMap& headers, bool) {
-  Formatter::HttpFormatterContext ctx{encoder_callbacks_->requestHeaders().ptr(), &headers};
-  config_->mutations().mutateResponseHeaders(headers, ctx, encoder_callbacks_->streamInfo());
+  Formatter::HttpFormatterContext context{encoder_callbacks_->requestHeaders().ptr(), &headers};
+  config_->mutations().mutateResponseHeaders(headers, context, encoder_callbacks_->streamInfo());
 
-  // If we haven't already traversed the route configs, do so now.
-  if (route_configs_.empty()) {
-    route_configs_ =
-        Http::Utility::getAllPerFilterConfig<PerRouteHeaderMutation>(encoder_callbacks_);
-  }
+  // Note if the filter before this one has send local reply then the decodeHeaders() will not be
+  // be called and the route config will not be initialized. Try it again here and it will be
+  // no-op if already initialized.
```

Comment:
Thanks for clarifying it here.
FWIW, I think that this makes filter implementations somewhat fragile. The implementation needs to take care of two states: the encode-part was invoked, and the encode-part wasn't invoked.
This is certainly not part of this PR, but I think a further consideration of whether the filter-manager should only invoke the sub-list of filters where the encode-part was invoked when there's a local-reply instead of the entire chain.

Reply:
> FWIW, I think that this makes filter implementations somewhat fragile.

🤣 Actually, our filter chain is not very robust. The problem we refered in our discussion is only a trivial one, the buffer processing, the async http call, there various details need to care to develop a bug free filter.

> This is certainly not part of this PR, but I think a further consideration of whether the filter-manager should only invoke the sub-list of filters where the encode-part was invoked when there's a local-reply instead of the entire chain.

This is not easy, esp considering that there some encoding-only filter.